src/StreamJsonRpc/MessageFormatterHelper.cs

```diff
@@ -0,0 +1,126 @@
+
+        /// <summary>
+        /// Object used to lock the acces to <see cref="requestProgressMap"/> and <see cref="progressMap"/>.
+        /// </summary>
+        protected readonly object progressLock = new object();
+
+        /// <summary>
+        /// Dictionary used to map the request id to their progress id token so that the progress objects are cleaned after getting the final response.
+        /// </summary>
+        protected readonly Dictionary<long, long> requestProgressMap = new Dictionary<long, long>();
+
+        /// <summary>
+        /// Dictionary used to map progress id token to its corresponding ProgressParamInformation instance containing the progress object and the necessary fields to report the results.
+        /// </summary>
+        protected readonly Dictionary<long, ProgressParamInformation> progressMap = new Dictionary<long, ProgressParamInformation>();
+
+        /// <summary>
+        /// Incrementable number to assing as token for the progress objects.
+        /// </summary>
+        protected long nextProgressId;
+
+        /// <summary>
+        /// Stores the id of the request currently being serialized so the converter can use it to create the request-progress map.
+        /// </summary>
+        protected long? requestIdBeingSerialized;
+
+        /// <summary>
+        /// Converts given <see cref="Type"/> to its <see cref="IProgress{T}"/> type.
+        /// </summary>
+        /// <param name="objectType">The type which may implement <see cref="IProgress{T}"/>.</param>
+        /// <returns>The <see cref="IProgress{T}"/> from given <see cref="Type"/> object, or <c>null</c>  if no such interface was found in the given <paramref name="objectType" />.</returns>
+        public static Type FindIProgressOfT(Type objectType)
+        {
+            Requires.NotNull(objectType, nameof(objectType));
+
+            if (objectType.IsConstructedGenericType && objectType.GetGenericTypeDefinition().Equals(typeof(IProgress<>)))
+            {
+                return objectType;
+            }
+
+            return objectType.GetTypeInfo().GetInterfaces().FirstOrDefault(i => i.IsConstructedGenericType && i.GetGenericTypeDefinition() == typeof(IProgress<>));
+        }
+
+        protected long AddProgressObjectToMap(object value)
+        {
+            if (this.requestIdBeingSerialized == null)
+            {
+                throw new NotSupportedException("IProgress<T> objects should not be part of any response or notification.");
+            }
+
+            lock (this.progressLock)
+            {
+                long progressId = this.nextProgressId++;
+                this.requestProgressMap.Add(this.requestIdBeingSerialized.Value, progressId);
+
+                this.progressMap.Add(progressId, new ProgressParamInformation(value));
+
+                return progressId;
+            }
+        }
+
+        protected class ProgressParamInformation
```

Comment:
>ProgressParamInformation [](start = 24, length = 24)

Please add xml doc comments to at least all the non-private members that explains things as if the consumer wasn't already familiar with them.